<a href="https://colab.research.google.com/github/hailes1/MCMProjects/blob/main/Final_Project_Moessner's_Theorem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Final Project - **Moessner's Theorem**
## Kylie Landa and Dagmawe Haileslassie

In this project we will be investigating and attempting to computationally prove Moessner's theorem. We will be utilizing Dexter Kozen and Alexandr Silva's paper on this theorem as a basis for our code and computations. 

**The basis of Moessner's theorem is as follows:**


To generate our first sequence we record every positive integer 1,2,3,... and we then cross out every nth element. From here we compute the prefix sums of the first sequence, ignoring the elements we have crossed out. From here we cross out every (n-1)st element. For the third and final sequence we compute the prefix sums of our second sequence and then we cross out every (n-2)nd element in our sequence. 


Moessner's theorem states that the final sequence we are left with will be 1^n, 2^n, 3^n,... 


We will begin our investigation by developing code that can carry out the process outlined above. We will then compare it with the sequence of Moessner's theorem to verify. 


After this, we will generate a generalization of Moessner's theorem known as Paasche's theorem. This theorem varies our parameters of the procedure (by increasing our step size by one each time). From this process we will prove that we can obtain a sequence of factorials as well as a sequence of super factorials. 

Finally, we will develop our own procedures based upon Moessner's theorem and investigate whether we are able to produce and relevant patterns and integer sequences using these new procedures. 



## **Investigation**

Begin by importing neccesary modules for this project. 

In [ ]:
import math
import random
import numpy as np
import matplotlib.pyplot as plt
from itertools import accumulate

We decided to code a function that would remove specified values from a list. We will see this come into play later. 

In [ ]:
def remove_values_from_list(the_list, val):
   return [value for value in the_list if value != val]

In [ ]:
list1 = [1,2,2,2,3,4,5]
remove_values_from_list(list1, 2)

[1, 3, 4, 5]

This function cumulativeSum(lst) returns our prefix sums that we will also utilize in the procedure for Moessner's theorem. 

In [ ]:
def cumulativeSum(lst): 
  return list(accumulate(lst))

## Original Moessner's Theorem 

When implementing this theorum we should note that the values either being deleted, skipped or crossed out as mentioned in the earlier part of this project are multiples of n at first, then multiples of n-1, multiples of n-2 etc... until the operation is no longer feasible. To do this let us first create a numsDeleted function that creates a list of numbers that are multiples of any value n we provide it. 

In [ ]:
def numsDeleted(numSteps, n):
  nums = [i*n for i in range(1,numSteps+1)]
  return nums

We begin by creating a function that will carry out our modifications to our starting sequence. A list of numbers with a specified value of n, excluding all multiples of n. Please note that we use numsDeleted in this function 

In [ ]:
def primaryList(numSteps, n):
  mainList = list(range(1,numSteps+1))
  nums = numsDeleted(numSteps, n)
  #print(nums)
  for i in range(numSteps):
    for j in range(numSteps):
      if mainList[i]==nums[j]:
        mainList[i] = 0
  mainList = remove_values_from_list(mainList, 0)
  return mainList

In [ ]:
examplelst = primaryList(25, 4)
examplelst

[1, 2, 3, 5, 6, 7, 9, 10, 11, 13, 14, 15, 17, 18, 19, 21, 22, 23, 25]

As we can see from the above code chunk, we have managed to create a list of numbers and successfully excluded multiples of n which in this case was 4. 

The next function will take in the original list of numbers from the primaryList function, find a cumilative sum for each values, cross out all multiples of n-1 and return a list of values. We decided to call this function normalMoessner. 

In [ ]:
def normalMoessner(lst, n):
  secondList = cumulativeSum(lst)
  if n > 2:
    for k in range(1, len(secondList)+1):
      if k%(n-1)==0:
        secondList[k-1]=0
    secondList = remove_values_from_list(secondList, 0)
  return secondList

In [ ]:
examplelst = primaryList(4, 4)
examplelst
examplelst2 = normalMoessner(examplelst, 4)
print(examplelst2)
examplelst3 = normalMoessner(examplelst2, 3)
print(examplelst3)
examplelst4 = normalMoessner(examplelst3, 2)
print(examplelst4)

[1, 3]
[1]
[1]


Looking at the above code chunk we can see that when running the normalMoessner function with decreasing values of n(n-1, n-2 ...) we reach closer and closer to the final sequence until are left with 1^n, 2^n, 3^n. This is what is shown in the list examplelst4. Instead of running the normalMoessner function repeatedly we thought it is important to implement a function that does those runs for us. That is what is implemented in the multipleRuns function. 

In [ ]:
def multipleRuns(numSteps, n):
  newlst = primaryList(numSteps, n)
  i = n
  while i>=2:
    newlst = normalMoessner(newlst, i)
    i -= 1
  return newlst

In [ ]:
print(multipleRuns(1000,13))
compare = [n**13 for n in range(1, 7)]
print(compare)

[1, 8192, 1594323, 67108864, 1220703125, 13060694016, 96889010407, 549755813888, 2541865828329, 10000000000000, 34522712143931, 106993205379072, 302875106592253, 793714773254144, 1946195068359375, 4503599627370496, 9904578032905937, 20822964865671168, 42052983462257059, 81920000000000000, 154472377739119461, 282810057883082752, 504036361936467383, 876488338465357824, 1490116119384765625, 2481152873203736576, 4052555153018976267, 6502111422497947648, 10260628712958602189, 15943230000000000000, 24417546297445042591, 36893488147419103232, 55040353993448503713, 81138303245565435904, 118272717781982421875, 170581728179578208256, 243569224216081305397, 344498040522809827328, 482880748567480579719, 671088640000000000000, 925103102315013629321, 1265437718438866624512, 1718264124282290785243, 2316779994178213904384, 3102863559971923828125, 4129065876983540801536, 5460999706120583177327, 7180192468708211294208, 9387480337647754305649, 12207031250000000000000, 15791096563156692195651, 20325604337

In the above code chunk we are able to see that he mutlipleRuns function for the value of 13 as n and 1000 numSteps seems to have  the same outcome as (n)^13, (n+1)^13 ... In the function comparewithGen we just want to have a return of true or false values if the above lists are the same. 

In [ ]:
def comparewithGen(numSteps, n):
  vals = multipleRuns(numSteps,n)
  compare = [x**n for x in range(1, numSteps)]
  for i in range(len(compare)):
    if vals[i]==compare[i]:
      return True
  return False

In [ ]:
iterate = [comparewithGen(1000, n) for n in range(4,200)]
print(iterate)

[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, Tru

From the above code chunk we can see that there are no false values present when comparing the results of the normal Moessner theorum function along side the compare list we created in comparewithGen function. This leads us to believe that we have successfully(through computation) proved that moessner's theorum is TRUE. 

## Increasing the constant n by 1

In the construction of Moessner’s theorem, the initial step size n is constant. What happens if we increase it in each step? Let us repeat the construction starting with a step size of one and increasing the step size by one each time.

We begin by creating a function that will carry out our modifications to our starting sequence. By creating a step variable that increases by 1 each time thorugh the while loop, we are increasing the index each step. We also increase i by each new step value and set the list value at this index to zero. Once this is completed we remove all zeros like we did above. This will return our first modified sequence. 

In [ ]:
def InclMoessner1(numSteps): 
  incList = list(range(1,numSteps+1))
  step = 2 
  i = 2 
  while i < len(incList): 
    incList[i] = 0 
    step += 1 
    i+= step 

  incList2 = remove_values_from_list(incList, 0)
  #print(step)
  return incList2

In [ ]:
InclMoessner1(21)

[1, 2, 4, 5, 7, 8, 9, 11, 12, 13, 14, 16, 17, 18, 19, 20]

Next, we call our function above to store this modified sequence and we begin our next step in the procedure. This is relatively similar, but we must calculate the prefix sums before we set any values to zero and remove them. Once we have calculated the prefix sums, the process is the same as above. 

In [ ]:
def InclMoessner2(numSteps): 
  incList2 = InclMoessner1(numSteps)
  #print(incList2)
  incList2.pop(0)
  incList3 = cumulativeSum(incList2)
  #print(incList3)
  step = 2 
  i = 2 
  while i < len(incList3): 
    incList3[i] = 0 
    step += 1 
    i+= step 
    
  incList3 = remove_values_from_list(incList3, 0)
  
  #print(step)
  return incList3

In [ ]:
InclMoessner2(21)

[2, 6, 18, 26, 46, 58, 71, 101, 118, 136, 155]

We continue this process and modify the sequence that we calculated with InclMoessner2 by calcuating the prefixes and then removing every n term (where n is increasing by 1 each time). 

In [ ]:
def InclMoessner3(numSteps): 
  incList3 = InclMoessner2(numSteps)
  #print(incList2)
  incList3.pop(0)
  incList4 = cumulativeSum(incList3)
  #print(incList3)
  step = 2 
  i = 2 
  while i < len(incList4): 
    incList4[i] = 0 
    step += 1 
    i+= step 

  incList4 = remove_values_from_list(incList4, 0)
  #print(step)
  return incList4

In [ ]:
InclMoessner3(21)

[6, 24, 96, 154, 326, 444, 580]

Once again, we do the same thing this time with the list returned by InclMoessner3 function above. 

In [ ]:
def InclMoessner4(numSteps): 
  incList4 = InclMoessner3(numSteps)
  #print(incList2)
  incList4.pop(0)
  incList5 = cumulativeSum(incList4)
  #print(incList3)
  step = 2 
  i = 2 
  while i < len(incList5): 
    incList5[i] = 0 
    step += 1 
    i+= step 

  incList5 = remove_values_from_list(incList5, 0)
  #print(step)
  return incList5

In [ ]:
InclMoessner4(21)

[24, 120, 600, 1044]

Lastly we do the same thing, but this time modifying our InclMoessner4 returned sequence. This will produce the final sequence, except for a few key terms which are the first terms of all of the previous sequences. 

In [ ]:
def InclMoessner5(numSteps): 
  incList5 = InclMoessner4(numSteps)
  #print(incList2)
  incList5.pop(0)
  incList6 = cumulativeSum(incList5)
  #print(incList3)
  step = 2 
  i = 2 
  while i < len(incList6): 
    incList6[i] = 0 
    step += 1 
    i+= step 

  incList6 = remove_values_from_list(incList6, 0)
  #print(step)
  return incList6

In [ ]:
InclMoessner5(50)

[120,
 720,
 4320,
 8028,
 22212,
 33984,
 48860,
 94852,
 128160,
 167868,
 214676,
 342964,
 427932,
 525240,
 635948,
 761166]

In order to account for the first terms of all of the produced sequences from each function, we append these first terms to our final InclMoessner5 list. This is our final function that we will call when testing this procedure. 

In [ ]:
def InclMoessner(numSteps): 
  lst1 = InclMoessner1(numSteps)
  lst2 = InclMoessner2(numSteps)
  lst3 = InclMoessner3(numSteps)
  lst4 = InclMoessner4(numSteps)

  terms1 = [lst1[0],lst2[0],lst3[0],lst4[0]]
  #print(terms1)

  terms2 = InclMoessner5(numSteps)
  result = [*terms1, *terms2] 

  return result 

In [ ]:
InclMoessner(50)

[1,
 2,
 6,
 24,
 120,
 720,
 4320,
 8028,
 22212,
 33984,
 48860,
 94852,
 128160,
 167868,
 214676,
 342964,
 427932,
 525240,
 635948,
 761166]

Let's compare this sequence with that of n!. We utilize the same code as above when we compared with 1^n, 2^n, 3^n..., but this time we modify our compare variables. We use a list comprehension to produce a list of n! values for n in range numSteps and compare this with InclMoessner(numSteps). If these values are consistent across indexes, the function will return True. 

In [ ]:
def comparewithFac(numSteps):
  vals = InclMoessner(numSteps)
  compare = [math.factorial(n) for n in range(1, numSteps)]
  for i in range(len(compare)):
    if vals[i]==compare[i]:
      return True
  return False

In [ ]:
comparewithFac(21)

True

In [ ]:
iterate2 = [comparewithFac(1000) for n in range(4,100)]
print(iterate2)

[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]


## What happens when we expand upon Koza and Silva's paper and develop our own procedure? 

Let's see what happens when we start with n=1 and increase it by 2 each time? 

### n = 1 and increase by 2 each step 

In order to implement this new procedure we simply modify our previous code that increases n each step. Instead of a step size of 1 we change the step size to be 2. We implement the same procedure as above. After each new sequence is produced, we enter it into the Encyclopedia of Integer sequences in order to understand if we have discovered a relevant pattern. 

In [ ]:
def Inclby2Moessner1(numSteps): 
  incList = list(range(2,numSteps+1))
  step = 2 
  i = 1
  while i < len(incList): 
    incList[i] = 0 
    step += 2 
    i+= step 

  incList2 = remove_values_from_list(incList, 0)
  print(step)
  return incList2

In [ ]:
Inclby2Moessner1(21)

10


[2, 4, 5, 6, 8, 9, 10, 11, 12, 14, 15, 16, 17, 18, 19, 20]

In [ ]:
def Inclby2Moessner2(numSteps): 
  incList2 = Inclby2Moessner1(numSteps)
  #print(incList2)
  incList2.pop(0)
  incList3 = cumulativeSum(incList2)
  #print(incList3)
  step = 2 
  i = 1
  while i < len(incList3): 
    incList3[i] = 0 
    step += 2 
    i+= step 
    
  incList3 = remove_values_from_list(incList3, 0)
  
  #print(step)
  return incList3

In [ ]:
print(Inclby2Moessner1(21))
Inclby2Moessner2(21)

10
[2, 4, 5, 6, 8, 9, 10, 11, 12, 14, 15, 16, 17, 18, 19, 20]
10


[4, 15, 23, 32, 53, 65, 79, 94, 110, 145, 164, 184]

In [ ]:
def Inclby2Moessner3(numSteps): 
  incList3 = Inclby2Moessner2(numSteps)
  #print(incList2)
  incList3.pop(0)
  incList4 = cumulativeSum(incList3)
  #print(incList3)
  step = 2 
  i = 1
  while i < len(incList4): 
    incList4[i] = 0 
    step += 2 
    i+= step 

  incList4 = remove_values_from_list(incList4, 0)
  #print(step)
  return incList4

In [ ]:
Inclby2Moessner3(21)

12


[15, 67, 107, 158, 234, 324, 550, 688, 845, 1022, 1220]

In [ ]:
def Inclby2Moessner4(numSteps): 
  incList4 = Inclby2Moessner3(numSteps)
  #print(incList2)
  incList4.pop(0)
  incList5 = cumulativeSum(incList4)
  #print(incList3)
  step = 2 
  i = 1
  while i < len(incList5): 
    incList5[i] = 0 
    step += 2
    i+= step 

  incList5 = remove_values_from_list(incList5, 0)
  #print(step)
  return incList5

In [ ]:
Inclby2Moessner4(21)

12


[67, 332, 566, 890, 1440, 2128, 3995, 5215]

In [ ]:
def Inclby2Moessner5(numSteps): 
  incList5 = Inclby2Moessner4(numSteps)
  #print(incList2)
  incList5.pop(0)
  incList6 = cumulativeSum(incList5)
  #print(incList3)
  step = 2 
  i = 1
  while i < len(incList6): 
    incList6[i] = 0 
    step += 2 
    i+= step 

  incList6 = remove_values_from_list(incList6, 0)
  #print(step)
  return incList6

In [ ]:
Inclby2Moessner5(21)

12


[332, 1788, 3228, 5356, 9351, 14566]

In [ ]:
def Incby2MoessnerConc(numSteps): 
  lst1 = Inclby2Moessner1(numSteps)
  lst2 = Inclby2Moessner2(numSteps)
  lst3 = Inclby2Moessner3(numSteps)
  lst4 = Inclby2Moessner4(numSteps)

  terms1 = [lst1[0],lst2[0],lst3[0],lst4[0]]
  #print(terms1)

  terms2 = Inclby2Moessner5(numSteps)
  result = [*terms1, *terms2] 

  return result 

In [ ]:
Incby2MoessnerConc(50)

18
18
18
18
18


[2,
 4,
 15,
 67,
 332,
 1788,
 3228,
 5356,
 9351,
 14566,
 29652,
 40082,
 53197,
 69370,
 89003,
 116887,
 149627,
 187755,
 282624,
 340683,
 406768,
 482434,
 568521,
 665912,
 775534,
 912584,
 1064926,
 1233677,
 1420004]

According to the on-line Encyclopedia of Integer Sequences, none of these sequences (for n = 2 with an increasing step size of 2) have been found significant before. In this regard, it seems we have not found an important pattern. Let's try a different approach. What if we keep n constant throughout the entire procedure and set it as n = 2? 

When changing our code to have an initial n value of 3 and an increase of 3 at each step size we found no significant integer sequences. The final sequence is attatched below. Feel free to modify the code to see this sequence develop. 

[2,
 4,
 15,
 67,
 332,
 1788,
 3228,
 5356,
 9351,
 14566,
 29652,
 40082,
 53197,
 69370,
 89003,
 116887,
 149627,
 187755,
 282624,
 340683,
 406768,
 482434,
 568521,
 665912,
 775534,
 912584,
 1064926,
 1233677,
 1420004]

## Keeping n constant as n = 2

We start by modifying our previous code. By changing our numsDeleted function, we create an array of factors of 2. 

In [ ]:
def numsDeleted2(numSteps):
  nums = [i*2 for i in range(1,numSteps+1)]
  return nums

We keep our primaryList function the same and make sure to accomodate our new numsDeleted function. This function removes all factors of 2 in our list. 

In [ ]:
def primaryList2(numSteps):
  mainList = list(range(1,numSteps+1))
  nums = numsDeleted2(numSteps)
  #print(nums)
  for i in range(numSteps):
    for j in range(numSteps):
      if mainList[i]==nums[j]:
        mainList[i] = 0
  mainList = remove_values_from_list(mainList, 0)
  return mainList

Next we calculate the prefix sums and from this list we remove any value that is divisible by 2. We then return this newly computed list. 

In [ ]:
def Moessnerby2(lst):
  secondList = cumulativeSum(lst)
  for k in range(1, len(secondList)+1):
    if k % (2)==0:
      secondList[k]=0
  secondList = remove_values_from_list(secondList, 0)
  return secondList

In [ ]:
testlst = primaryList2(50)
print(testlst)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49]


In [ ]:
testlst2by2 = Moessnerby2(testlst)
print(testlst2by2)

[1, 4, 16, 36, 64, 100, 144, 196, 256, 324, 400, 484, 576]


Our new modified list appears to exhibit a newfound pattern. This list includes the squared value of the even number sequence (2,4,6,8,...) and 1. When we search this sequence on the Encyclopedia of Integer Sequences  we find a recorded entry (#A055808). 

# **Discussion** 

## Conclusion and Conjectures 
In this project we successfully investigated and computationally proved Moessner's theorem. We utilized Dexter Kozen and Alexandr Silva's paper on this theorem as a basis for our code and computations. The team was able to see that with crossing out decreasing values of n(n-1, n-2 ...) for mutliple sequences we reached the final sequence of 1^n, 2^n, 3^n after each run. 

We then successfully investigated and computationally proved Moessner's theorum but this time we increased the constant n by 1 each time. (Refer to the third section)

Part the project requirments was to come up with our own variant
of Moessner’s construction and look for numerical patterns. We were able to implement this in two different ways. First we wanted to see what happens when we increase the constant n by 2 each time. We unfortunatley were not able to see any significant sequences for this test. The second test we tried to do is keep n constant (which in our case is 2). Our new modified list appears to exhibit a newfound pattern. This list includes the squared value of the even number sequence (2,4,6,8,...) and When we search this sequence on the Encyclopedia of Integer Sequences we find a recorded entry (#A055808). 

## Limitations 
One of the major limitations when working with this project is the time it takes to find a new pattern when looking at different variants of moessners construction. Are there more patterns if we cross out prime numbers and continue doing that for every sequence? Are there more patterns when n is constant at 3, or constant at 4? What would be the True/False values for when n>200?